src/middlewared/middlewared/plugins/apps/resources.py

```diff
@@ -142,3 +142,20 @@ async def gpu_choices_internal(self):
             await self.middleware.call('device.get_gpus'),
             await self.middleware.run_in_thread(get_nvidia_gpus),
         )
+
+    @private
+    async def get_hostpaths_datasets(self, app_name):
+        app_info = await self.middleware.call('app.get_instance', app_name)
+        host_paths = [
+            volume['source'] for volume in app_info['active_workloads']['volumes']
+            if volume['source'].startswith(f'{IX_APPS_MOUNT_PATH}/') is False
+        ]
+
+        mapping = {}
+        for host_path in host_paths:
+            try:
+                mapping[host_path] = await self.middleware.call('zfs.dataset.path_to_dataset', host_path)
```

Comment:
This is unnecessarily expensive and I'm not keen on it. The `path_to_dataset` could be moved to a utility function so that we're not using a child process for each host path. Once `path_to_dataset` was in its own function, you could put the for loop into its own thread.

```
mapping = await self.middleware.run_in_thread(path_to_dataset, host_paths)